core/src/main/java/hudson/PluginManager.java

```diff
@@ -959,12 +960,35 @@ public HttpResponse doUploadPlugin(StaplerRequest req) throws IOException, Servl
 
             pluginUploaded = true;
 
+            JSONArray dependencies = new JSONArray();
+            try {
+                Manifest m = new JarFile(t).getManifest();
+                String deps = m.getMainAttributes().getValue("Plugin-Dependencies");
+                // create the dependencies object, so that the "download" can fail and people will know what
+                // dependencies are needed for the upload.
+
+                if (StringUtils.isNotBlank(deps)) {
+                    // now we get to parse it!
+                    String[] plugins = deps.split(",");
+                    for (String p : plugins) {
+                        // should have name:version[;resolution:=optional]
+                        String[] attrs = p.split("[:;]");
```

Comment:
Yes, I see the PluginWrapper.Depedency class that could be used. I'm not sure it would make it much simpler though.